File: lru/lru.ts
// import { v4 as uuid4 } from 'uuid4'
let id = 0

export type LRUNode<T> = {
  value: T;
  left?: LRUNode<T>;
  right?: LRUNode<T>;
}

class LRU<T> {
	public cache: { [id: string]: LRUNode<T> }
	public head :LRUNode<T> | undefined
	public tail : LRUNode<T> | undefined
  constructor() {
     this.cache = {},
	 this.head = undefined,
	 this.tail = undefined
  }

  update(k: number) {
    const node = this.cache[k];
	if (!node){
		return
	}
	if (node == this.tail){
		this.tail = node.left!
	}
    if (node.left) {
      const prev = node.left;
      const next = node.right;
      prev.right = next;
	}
	node.right = this.head  
	this.head = node
	node.left = undefined
  }

  remove():T {
		const tail = this.tail!
		this.tail = tail.left!
		this.tail.right = undefined
		tail.left = undefined
		return tail.value
  }

  append(value: T){
		const newNode: LRUNode<T> = { value : value } 
		this.cache[id++] = newNode
		if (!this.head){
			this.head = newNode
			return
		} else if (!this.tail) {
			this.head.right = newNode
			newNode.left = this.head
		} else {
			this.tail.right = newNode
			newNode.left = this.tail
		}
		this.tail = newNode
	}

  display(node: LRUNode<T> | undefined = undefined){
		if (!node){
			node = this.head!
		}
		console.log(node!.value, ' -> ')
		if(node.right){
			this.display(node.right)
		}
  }
}


const lru = new LRU()
lru.append('foo')
lru.append('bar')
lru.append('fiz')
lru.append('buzz')
lru.append('fizzbuzz')

lru.display()
console.log(' ')
lru.update(4)
lru.display()
console.log(' ')
lru.update(1)
lru.display()
console.log(' ')
console.log(lru.remove(), 'removed')
lru.display()

File: lru/LRU.py
from uuid import uuid4

class Node:

	def __init__(self, value):
		self.value = value
		self.next = None
		self.prev = None

class LeastRecentlyUsed:

	def __init__(self, cap):
		self.cap = cap
		self.head = None
		self.tail = None
		self.length = 0
		self.cache = {}
		self.reverse_cache = {}

	def get(self, v):
		if v in self.reverse_cache:
			id = self.reverse_cache[v]
			node = self.cache[id]
		else:
			node = Node(v)
			self.reverse_cache[node] = uuid4()
			id = self.reverse_cache[node]
			self.cache[id] = node
			self.length += 1
		self.detach(node)
		self.prepend(node)
		self.trim_cache()
		
	def put(self, id, new_value):
		pass

	def detach(self, node):
		prev = node.left
		next = node.right
		if prev:
			prev.next = next
		node.left = None
		node.right = None
		if node == self.tail:
			self.tail = prev

	def prepend(self, node):
		node.right = self.head
		if not self.tail and node.right:
			self.tail = node.right
		if self.head:
			self.head.left = node

	def trim_cache(self):
		if self.length > self.cap:
			tail = self.tail
			self.tail = tail.prev
			tail.left = tail.right = None
		return tail.value
	